File: backend/src/bond_monitor/domain/bonds/offers.py
"""Bond offer types and read-model helpers (put / call)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from enum import StrEnum
from typing import Protocol

from bond_monitor.domain.shared.formatting import format_date


class OfferKind(StrEnum):
    PUT = "put"
    CALL = "call"


class OfferWindowStatus(StrEnum):
    """Lifecycle of an investor put-offer submission window."""

    UNKNOWN = "unknown"
    NOT_OPEN = "not_open"
    OPEN = "open"
    CLOSED = "closed"
    EXPIRED = "expired"


class PutOfferDecision(StrEnum):
    PENDING = "pending"
    EXERCISE = "exercise"
    HOLD = "hold"


PUT_OFFER_DECISION_LABELS: dict[PutOfferDecision, str] = {
    PutOfferDecision.PENDING: "Не решено",
    PutOfferDecision.EXERCISE: "Предъявить",
    PutOfferDecision.HOLD: "Держать до погашения",
}

OFFER_WINDOW_STATUS_LABELS: dict[OfferWindowStatus, str] = {
    OfferWindowStatus.UNKNOWN: "Окно подачи не объявлено",
    OfferWindowStatus.NOT_OPEN: "Приём заявок ещё не начался",
    OfferWindowStatus.OPEN: "Можно подать заявку",
    OfferWindowStatus.CLOSED: "Окно подачи закрыто",
    OfferWindowStatus.EXPIRED: "Оферта прошла",
}


class _OfferScheduleFields(Protocol):
    offer_date: date | None
    offer_submission_start: date | None
    offer_submission_end: date | None
    offer_price_pct: float | None
    call_date: date | None


@dataclass(frozen=True)
class BondOfferView:
    """Nearest investor put-offer snapshot for API/UI."""

    kind: OfferKind
    execution_date: date
    submission_start: date | None
    submission_end: date | None
    price_pct: float | None
    window_status: OfferWindowStatus
    moex_offer_type: str | None = None


def offer_window_status(
    *,
    offer_date: date | None,
    submission_start: date | None,
    submission_end: date | None,
    as_of: date,
) -> OfferWindowStatus | None:
    """Return window status for a put-offer, or ``None`` if there is no future offer."""
    if offer_date is None:
        return None
    if offer_date <= as_of:
        return OfferWindowStatus.EXPIRED
    if submission_start is None and submission_end is None:
        return OfferWindowStatus.UNKNOWN
    if submission_start is not None and as_of < submission_start:
        return OfferWindowStatus.NOT_OPEN
    if submission_end is not None and as_of > submission_end:
        return OfferWindowStatus.CLOSED
    return OfferWindowStatus.OPEN


def bond_offer_view(
    source: _OfferScheduleFields,
    as_of: date,
    *,
    kind: OfferKind = OfferKind.PUT,
    moex_offer_type: str | None = None,
) -> BondOfferView | None:
    """Build put-offer read-model from bond or position fields."""
    if source.offer_date is None:
        return None
    window_status = offer_window_status(
        offer_date=source.offer_date,
        submission_start=source.offer_submission_start,
        submission_end=source.offer_submission_end,
        as_of=as_of,
    )
    if window_status is None:
        return None
    return BondOfferView(
        kind=kind,
        execution_date=source.offer_date,
        submission_start=source.offer_submission_start,
        submission_end=source.offer_submission_end,
        price_pct=source.offer_price_pct,
        window_status=window_status,
        moex_offer_type=moex_offer_type,
    )


def put_offer_awareness_message(view: BondOfferView) -> str:
    """Human-readable informational text for watch reminders."""
    execution = format_date(view.execution_date)
    if view.window_status == OfferWindowStatus.UNKNOWN:
        return f"Пут-оферта {execution} — окно подачи ещё не объявлено эмитентом"
    if view.window_status == OfferWindowStatus.NOT_OPEN and view.submission_start is not None:
        return (
            f"Пут-оферта {execution} — приём заявок с "
            f"{format_date(view.submission_start)}"
        )
    if view.window_status == OfferWindowStatus.OPEN and view.submission_end is not None:
        return (
            f"Пут-оферта {execution} — подайте заявку до "
            f"{format_date(view.submission_end)} включительно"
        )
    return f"Пут-оферта {execution}"


def put_offer_action_message(view: BondOfferView) -> str:
    """Human-readable action text when submission window is open."""
    if view.submission_end is not None:
        return (
            f"Подайте заявку на пут-оферту до "
            f"{format_date(view.submission_end)} включительно "
            f"(исполнение {format_date(view.execution_date)})"
        )
    return f"Подайте заявку на пут-оферту (исполнение {format_date(view.execution_date)})"

File: backend/internal/domain/bonds/offers.go
package bonds

import (
	"fmt"
	"time"

	"github.com/tonatos/bond-monitor/backend/internal/domain/shared"
)

type OfferKind string

const OfferKindPut OfferKind = "put"

type OfferWindowStatus string

const (
	OfferWindowUnknown  OfferWindowStatus = "unknown"
	OfferWindowNotOpen  OfferWindowStatus = "not_open"
	OfferWindowOpen     OfferWindowStatus = "open"
	OfferWindowClosed   OfferWindowStatus = "closed"
	OfferWindowExpired  OfferWindowStatus = "expired"
)

type PutOfferDecision string

const (
	PutOfferPending  PutOfferDecision = "pending"
	PutOfferExercise PutOfferDecision = "exercise"
	PutOfferHold     PutOfferDecision = "hold"
)

type OfferSchedule interface {
	GetOfferDate() *time.Time
	GetOfferSubmissionStart() *time.Time
	GetOfferSubmissionEnd() *time.Time
	GetOfferPricePct() *float64
	GetCallDate() *time.Time
}

type BondOfferView struct {
	Kind            OfferKind
	ExecutionDate   time.Time
	SubmissionStart *time.Time
	SubmissionEnd   *time.Time
	PricePct        *float64
	WindowStatus    OfferWindowStatus
	MoexOfferType   *string
}

func OfferWindowStatusFor(
	offerDate, submissionStart, submissionEnd *time.Time,
	asOf time.Time,
) *OfferWindowStatus {
	if offerDate == nil {
		return nil
	}
	asOf = shared.DateOnly(asOf)
	od := shared.DateOnly(*offerDate)
	if !od.After(asOf) {
		s := OfferWindowExpired
		return &s
	}
	// Incomplete window (only start or only end) is not actionable — wait for both dates.
	if submissionStart == nil || submissionEnd == nil {
		s := OfferWindowUnknown
		return &s
	}
	if asOf.Before(shared.DateOnly(*submissionStart)) {
		s := OfferWindowNotOpen
		return &s
	}
	if asOf.After(shared.DateOnly(*submissionEnd)) {
		s := OfferWindowClosed
		return &s
	}
	s := OfferWindowOpen
	return &s
}

func BondOfferViewFrom(source OfferSchedule, asOf time.Time) *BondOfferView {
	offerDate := source.GetOfferDate()
	if offerDate == nil {
		return nil
	}
	status := OfferWindowStatusFor(
		offerDate,
		source.GetOfferSubmissionStart(),
		source.GetOfferSubmissionEnd(),
		asOf,
	)
	if status == nil {
		return nil
	}
	return &BondOfferView{
		Kind:            OfferKindPut,
		ExecutionDate:   shared.DateOnly(*offerDate),
		SubmissionStart: source.GetOfferSubmissionStart(),
		SubmissionEnd:   source.GetOfferSubmissionEnd(),
		PricePct:        source.GetOfferPricePct(),
		WindowStatus:    *status,
	}
}

func PutOfferAwarenessMessage(view BondOfferView) string {
	execution := shared.FormatDate(&view.ExecutionDate)
	switch view.WindowStatus {
	case OfferWindowUnknown:
		return fmt.Sprintf("Пут-оферта %s — окно подачи ещё не объявлено эмитентом", execution)
	case OfferWindowNotOpen:
		if view.SubmissionStart != nil {
			return fmt.Sprintf(
				"Пут-оферта %s — приём заявок с %s",
				execution,
				shared.FormatDate(view.SubmissionStart),
			)
		}
	case OfferWindowOpen:
		if view.SubmissionEnd != nil {
			return fmt.Sprintf(
				"Пут-оферта %s — подайте заявку до %s включительно",
				execution,
				shared.FormatDate(view.SubmissionEnd),
			)
		}
	}
	return fmt.Sprintf("Пут-оферта %s", execution)
}

func PutOfferActionMessage(view BondOfferView) string {
	if view.SubmissionEnd != nil {
		return fmt.Sprintf(
			"Подайте заявку на пут-оферту до %s включительно (исполнение %s)",
			shared.FormatDate(view.SubmissionEnd),
			shared.FormatDate(&view.ExecutionDate),
		)
	}
	return fmt.Sprintf(
		"Подайте заявку на пут-оферту (исполнение %s)",
		shared.FormatDate(&view.ExecutionDate),
	)
}

// OfferWindowData is put-offer schedule fields from an external source (MOEX / T-Invest).
type OfferWindowData struct {
	OfferDate       *time.Time
	SubmissionStart *time.Time
	SubmissionEnd   *time.Time
	PricePct        *float64
}

// SelectOfferWindow picks the offer matching preferDate, else the nearest future offer on/after asOf.
func SelectOfferWindow(offers []OfferWindowData, preferDate *time.Time, asOf time.Time) *OfferWindowData {
	if len(offers) == 0 {
		return nil
	}
	asOf = shared.DateOnly(asOf)
	if preferDate != nil {
		prefer := shared.DateOnly(*preferDate)
		for i := range offers {
			if offers[i].OfferDate == nil {
				continue
			}
			if shared.DateOnly(*offers[i].OfferDate).Equal(prefer) {
				return &offers[i]
			}
		}
	}
	var best *OfferWindowData
	for i := range offers {
		od := offers[i].OfferDate
		if od == nil {
			continue
		}
		d := shared.DateOnly(*od)
		if d.Before(asOf) {
			continue
		}
		if best == nil || d.Before(shared.DateOnly(*best.OfferDate)) {
			best = &offers[i]
		}
	}
	return best
}

// ApplyOfferWindow fills missing put-offer window fields. Does not overwrite set values.
// When bond.OfferDate and patch.OfferDate are both set, they must match.
func ApplyOfferWindow(bond *BondRecord, patch OfferWindowData) bool {
	if bond == nil {
		return false
	}
	if bond.OfferDate != nil && patch.OfferDate != nil {
		if !shared.DateOnly(*bond.OfferDate).Equal(shared.DateOnly(*patch.OfferDate)) {
			return false
		}
	}
	changed := false
	if bond.OfferDate == nil && patch.OfferDate != nil {
		d := shared.DateOnly(*patch.OfferDate)
		bond.OfferDate = &d
		changed = true
	}
	if bond.OfferSubmissionStart == nil && patch.SubmissionStart != nil {
		d := shared.DateOnly(*patch.SubmissionStart)
		bond.OfferSubmissionStart = &d
		changed = true
	}
	if bond.OfferSubmissionEnd == nil && patch.SubmissionEnd != nil {
		d := shared.DateOnly(*patch.SubmissionEnd)
		bond.OfferSubmissionEnd = &d
		changed = true
	}
	if bond.OfferPricePct == nil && patch.PricePct != nil {
		v := *patch.PricePct
		bond.OfferPricePct = &v
		changed = true
	}
	return changed
}
